### PR TITLE
Query result validator

### PR DESCRIPTION
**Pull request by Natoandro**
- Compiled validator
- Only include selected fields in the result (no additional step for removal)
- Selections for union of object types

### CODE REVIEW

File: typegate/src/engine.ts [hunk 1]
```diff
@@ -285,14 +339,45 @@ export class Engine {
...
+          return variant;
+        });
+        for (const variant of distinct(resultVariants)) {
+          cache[`${stageId}$${variant}`] = [];
+          lenses[`${stageId}$${variant}`] = [];
+        }
+        for (const [i, variant] of resultVariants.entries()) {
+          cache[`${stageId}$${variant}`].push(cache[stageId][i]);
+          lenses[`${stageId}$${variant}`].push(lenses[stageId][i]);
+        }
```
Comment (afmika) [inline typegate/src/engine.ts:377]: I would suggest to refactor `(stageId, variant) -> string` to a function

File: typegate/src/typecheck/result.ts [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+              cg.generateStringValidator(typeNode);
+              break;
+          }
+        }
+      } else {
+        // TODO: cannot check enum - perhaps we should disable enums for non-scalar types??
+
+        switch (typeNode.type) {
+          case "optional": {
+            const c = entry.selectionSet != null ? ++this.counter : null;
```
Comment (afmika) [inline typegate/src/typecheck/result.ts:129]: I think that's a good idea.

File: typegate/src/typecheck/common.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+      return true;
+    } catch (_e) {
+      return false;
+    }
+  },
+  email: validator.isEmail,
+  // TODO validatorjs does not have a URI validator, so this is stricter than expected
+  uri: (value: string) => {
+    console.log("validating uri", value);
+    return validator.isURL(value, {
```
Comment (afmika) [inline typegate/src/typecheck/common.ts:35]:
```typescript
return validator.isDataURI(value) || validator.isURL(value, ...)
```

**Review by afmika — COMMENTED** [review]
Nice 👍

File: typegate/src/engine.ts [hunk 4]
```diff
@@ -219,11 +235,14 @@ export class Engine {
     verbose: boolean,
   ): Promise<JSONValue> {
     const ret = {};
-    const cache: Record<string, unknown> = {};
-    const lenses: Record<string, unknown> = {};
+    const cache: Record<string, unknown[]> = {};
```
Comment (zifeo) [inline typegate/src/engine.ts:238]: why is an array required? maybe you can add comment on the example data cached and put in the lenses?

File: typegate/src/engine.ts [hunk 5]
```diff
@@ -236,16 +255,51 @@ export class Engine {
...
       } = stage.props;
 
+      const level = path.length;
+      const stageId = stage.id();
+      // const parentId = parent.id();
+      const parentId = parent && stageId.slice(0, stageId.lastIndexOf("."));
+
+      let activeSelection = getActiveSelection();
+      while (activeSelection != null && level < activeSelection.level) {
+        selections.pop();
```
Comment (zifeo) [inline typegate/src/engine.ts:268]: Add a test for the while loop?

File: typegate/src/engine.ts [hunk 6]
```diff
@@ -236,16 +255,51 @@ export class Engine {
...
+      }
+
+      if (activeSelection != null) {
+        if (parentId == null) {
+          throw new Error("unexpected state");
+        }
+
+        if (activeSelection.level === level) {
+          // potential variant transition
+          const match = parentId?.match(VARIANT_TRANSITION_REGEX);
```
Comment (zifeo) [inline typegate/src/engine.ts:279]: regex matching is costly, is there another alternative?

File: typegate/src/engine.ts [hunk 7]
```diff
@@ -285,14 +339,45 @@ export class Engine {
...
       }
-      const field = path[path.length - 1] as any;
+
+      const field = path[path.length - 1];
       if (node !== "") {
         lens.forEach((l: any, i: number) => {
-          l[field] = res[i];
+          // Objects are replaced by empty objects `{}`.
+          // It will be populated by child compute stages using values in `cache`.
+          l[field] = withEmptyObjects(res[i]);
```
Comment (zifeo) [inline typegate/src/engine.ts:348]: if it is replaced anyway later, why putting empty object?

File: typegate/src/planner/mod.ts [hunk 8]
```diff
@@ -109,90 +105,138 @@ export class Planner {
...
-        selection.map((n) => n.name?.value),
-        typ.type,
-        Object.entries(props).reduce(
-          (agg, [k, v]) => ({ ...agg, [k]: this.tg.type(v).type }),
-          {},
-        ),
-      );
+    if (selectionSet == null) {
+      if (this.isSelectionSetExpectedFor(typeIdx)) {
+        const path = this.formatPath(node.path);
```
Comment (zifeo) [inline typegate/src/planner/mod.ts:111]: add a test for here?

File: typegate/src/planner/mod.ts [hunk 9]
```diff
@@ -109,90 +105,138 @@ export class Planner {
...
+      const selections = selectionSet.selections.map((sel) => {
+        if (sel.kind !== Kind.INLINE_FRAGMENT || sel.typeCondition == null) {
+          throw new Error("Expected inline fragment with type condition");
         }
-        // if the above fails, tell the user in case they made a typo
-        const suggestion = closestWord(name, allProps);
-        if (suggestion) {
+        const typeName = sel.typeCondition.name.value;
+        if (!unselectedVariants.has(typeName)) {
+          const path = this.formatPath(node.path);
```
Comment (zifeo) [inline typegate/src/planner/mod.ts:160]: test for here?

File: typegate/src/planner/mod.ts [hunk 10]
```diff
@@ -501,4 +504,73 @@ export class Planner {
   private formatPath(path: string[]) {
     return [this.operationName, ...path].join(".");
   }
+
+  private isSelectionSetExpectedFor(typeIdx: number): boolean {
+    const typ = this.tg.type(typeIdx);
+    if (typ.type === Type.OBJECT) {
+      return true;
```
Comment (zifeo) [inline typegate/src/planner/mod.ts:511]: test for the next three code cov warning?

File: typegate/src/typecheck/matching_variant.ts [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+        continue;
+      }
+      const typeNode = this.tg.type(typeIdx);
+
+      if (typeNode.enum != null) {
+        cg.generateEnumValidator(typeNode);
+      } else {
+        switch (typeNode.type) {
+          case "boolean":
+            cg.generateBooleanValidator(typeNode);
```
Comment (zifeo) [inline typegate/src/typecheck/matching_variant.ts:80]: test?

File: typegate/src/typecheck/matching_variant.ts [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+          case "object":
+            cg.generateObjectValidator(
+              typeNode,
+              mapValues(typeNode.properties, functionName),
+            );
+            queue.push(...Object.values(typeNode.properties));
+            break;
+          case "union":
+            cg.generateUnionValidator(
+              typeNode,
```
Comment (zifeo) [inline typegate/src/typecheck/matching_variant.ts:106]: test for the next warnings?

File: typegate/src/typecheck/result.ts (hunk 2 above)
Comment (zifeo) [reply typegate/src/typecheck/result.ts:129]: can we avoid ++ and use += 1 for clarity?

File: typegate/src/typecheck/result.ts [hunk 13]
```diff
@@ -0,0 +1,387 @@
...
+      }
+
+      case Kind.INLINE_FRAGMENT: {
+        if (node.typeCondition != null) {
+          throw new Error("Unexpected type condition on non-union type");
+        }
+        return node.selectionSet.selections.flatMap((selectionNode) =>
+          this.getChildEntriesFromSelectionNode(
+            typeNode,
+            entry,
```
Comment (zifeo) [inline typegate/src/typecheck/result.ts:290]: test?

**Review by zifeo — COMMENTED** [review]
very nice!

File: typegate/src/typecheck/result.ts (hunk 2 above)
Comment (Natoandro) [reply typegate/src/typecheck/result.ts:129]: > I don't quite follow, aren't `t.enum` variants `str` only ?

`t.enum()` is an alias to `t.string().enum()`. Any type is valid.

File: typegate/src/engine.ts (hunk 7 above)
Comment (Natoandro) [reply typegate/src/engine.ts:348]: Objects always start empty. The fields are set by child compute stages (corresponding to the field selections in the query).

Comment (zifeo) [reply typegate/src/engine.ts:348]: Yes, but letting `l[field] = res[I]` would avoid creating new object and still do the same behaviour? In this part of the engine, we should avoid creating or manipulating object/values so we have the less overhead possible.

Comment (Natoandro) [reply typegate/src/engine.ts:348]: The object might have fields that are not  selected by the query. Replacing with empty object will ensure that we only have the selected fields in the final result.

**Review by zifeo — APPROVED** [review]

**Pull request merged**